### ISSUE TITLE
chunks not match

### ISSUE THREAD

**Issue by lizhimao** [mravanelli/pytorch-kaldi#109]
Hi, dear developers, i'm a student working with the pytorch_kaldi. I have a wired question with the number of chunks. In the `.cfg` file i set the n_chunks(for train set) to 50, but in terminal i saw this:

```
------------------------------ Epoch 00 / 23 ------------------------------
Training tedlium_tr chunk = 1 / 51

```
and after running the training 50/51, the program will be killed normally. I saw feature path in `train_tedlium_tr_ep00_ck50.cfg` is the same as original configuration file, i think this is the problem why the program is killed. (If i set chunks = 50, it should not have `*_ck50.cfg` right?)

and in `list_chunks.txt` the valid and test set have the same chunks as the train set, although i set the `n_chunks =1` for valid and test.

I think change `compute_n_chunks` in `utilts.py` is not the right way to solve this problem. I think i'm using the right features and alignments. Can you help me? Thank you very much!

**Comment by mravanelli** [comment]
Hi,
thank you for your interest into the pytorch-kaldi project.
Could you send us the config file?

Mirco

**Comment by lizhimao** [comment]
Hi,
thank you for the quickly response, this is my config file
```
[cfg_proto]
cfg_proto = proto/global.proto
cfg_proto_chunk = proto/global_chunk.proto

[exp]
cmd =
run_nn_script = run_nn
out_folder = exp/tedlium_MLP_mfcc_ep24_ck50_standard
seed = 1234
use_cuda = True
multi_gpu = False
save_gpumem = False
n_epochs_tr = 24

[dataset1]
data_name = tedlium_tr
fea = fea_name=mfcc
	fea_lst=/home/kaldi/egs/tedlium/s5_r2/data/train/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/kaldi/egs/tedlium/s5_r2/data/train/utt2spk  ark:/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/train_cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=2 ark:- ark:- |
	cw_left=5
	cw_right=5


lab = lab_name=lab_cd
	lab_folder=/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/ali_train
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/kaldi/egs/tedlium/s5_r2/data/train/
	lab_graph=/home/kaldi/egs/tedlium/s5_r2/exp/tri3/graph


n_chunks = 50

[dataset2]
data_name = tedlium_dev
fea = fea_name=mfcc
	fea_lst=/home/kaldi/egs/tedlium/s5_r2/data/dev/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/kaldi/egs/tedlium/s5_r2/data/dev/utt2spk  ark:/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/dev_cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=2 ark:- ark:- |
	cw_left=5
	cw_right=5


lab = lab_name=lab_cd
	lab_folder=/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/ali_dev
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/kaldi/egs/tedlium/s5_r2/data/dev/
	lab_graph=/home/kaldi/egs/tedlium/s5_r2/exp/tri3/graph


n_chunks = 1

[dataset3]
data_name = tedlium_test
fea = fea_name=mfcc
	fea_lst=/home/kaldi/egs/tedlium/s5_r2/data/test/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/zhiyi/kaldi/egs/tedlium/s5_r2/data/test/utt2spk  ark:/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/test_cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=2 ark:- ark:- |
	cw_left=5
	cw_right=5


lab = lab_name=lab_cd
	lab_folder=/home/kaldi/egs/tedlium/s5_r2/exp/tedlium/ali_test
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/kaldi/egs/tedlium/s5_r2/data/test/
	lab_graph=/home/kaldi/egs/tedlium/s5_r2/exp/tri3/graph


n_chunks = 1

[data_use]
train_with = tedlium_tr
valid_with = tedlium_dev
forward_with = tedlium_test

[batches]
batch_size_train = 128
max_seq_length_train = 1000
increase_seq_length_train = False
start_seq_len_train = 100
multply_factor_seq_len_train = 2
batch_size_valid = 128
max_seq_length_valid = 1000

[architecture1]
arch_name = MLP_layers1
arch_proto = proto/MLP.proto
arch_library = neural_networks
arch_class = MLP
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = False
dnn_lay = N_out_lab_cd
dnn_drop = 0.2
dnn_use_laynorm_inp = False
dnn_use_batchnorm_inp = False
dnn_use_batchnorm = True
dnn_use_laynorm = False
dnn_act = softmax
arch_lr = 0.08
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = adam
opt_betas = 0.9,0.999
opt_weight_decay = 0.0
opt_eps = 1e-8
opt_amsgrad = False

[model]
model_proto = proto/model.proto
model = out_dnn1=compute(MLP_layers1,mfcc)
	loss_final=cost_nll(out_dnn1,lab_cd)
	err_final=cost_err(out_dnn1,lab_cd)

[forward]
forward_out = out_dnn1
normalize_posteriors = True
normalize_with_counts_from = lab_cd
save_out_file = False
require_decoding = True

[decoding]
decoding_script_folder = kaldi_decoding_scripts/
decoding_script = decode_dnn.sh
decoding_proto = proto/decoding.proto
min_active = 200
max_active = 7000
max_mem = 50000000
beam = 13.0
latbeam = 8.0
acwt = 0.2
max_arcs = -1
skip_scoring = false
scoring_script = local/score.sh
scoring_opts = "--min-lmwt 1 --max-lmwt 10"
norm_vars = False

```

**Comment by TParcollet** [comment]
Hi ! Could you please first delete the 'exp/tedlium_MLP_mfcc_ep24_ck50_standard' folder, and launch the experiment again ? SO we can verify this error isn't due to a weird previous launch ?

**Comment by lizhimao** [comment]
I tried this. But it didn't work :(

**Comment by mravanelli** [comment]
That is weird, we are not able to replicate it, right Titouan?

**Comment by TParcollet** [comment]
Indeed ... I tried with the same configuration, and I obtain 50 chunks not 51. Could you please paste the content of /exp_files/list_chunks.txt

**Comment by lizhimao** [comment]
Yes, hier is the list_chunks.txt

[list_chunks.txt](https://github.com/mravanelli/pytorch-kaldi/files/3278154/list_chunks.txt)

**Comment by TParcollet** [comment]
Do you have the latest version of Pytorch-Kaldi ?

**Comment by lizhimao** [comment]
Yes, I tried it with the latest version of pytorch-kaldi.

**Comment by TParcollet** [comment]
@lizhimao could you please try to change train_with = tedlium_tr by train_with = tedlium_dev and set the number of chunks to 2 instead of one ?

**Comment by lizhimao** [comment]
@TParcollet thank you very much, I will try. By the way, is the DNN training with kaldi necessary, I mean this:
```
cd kaldi/egs/timit/s5
./local/nnet/run_dnn.sh
```
If I only use tri3 alignments?

**Comment by lizhimao** [comment]
@TParcollet I tried your method and it works. After that I set the parameter as before, and the chunks is right... the list_chunks.txt is also right, only have one dev every epoch. I'm happy it works, but can't figure out why...

**Comment by TParcollet** [comment]
No the nnet step is not necessary, unless you want to train from the alignements generated by nnet. Overall, it's just to compare. Could you please describe what you did ? You just swapped the train and the dev, ran once, and then came back to the exact same cfg file and everything worked ?

**Comment by lizhimao** [comment]
First I changed 'train_with = tedlium_tr' by 'train_with = tedlium_dev' and N_chunks = 2(in dev), and run the script with this configuration, in terminal shows 'Training tedlium_dev chunk = 1 / 2'.

Then, I delete the exp file in 'out_folder', and set the 'train_with' back to 'tedlium_tr', the N_chunks in dev to 1. Start to run the script with this config(the original config). In terminal I can see 'Training tedlium_tr chunk = 1 / 50'. 

Actully the same as you said.

**Comment by TParcollet** [comment]
Okay great, maybe just an encoding problem of a weird character in the cfg file. !

**Comment by lizhimao** [comment]
I tried again with other configs, before run, I deleted the exp file. With the way can not solve the problem actually...The problem comes again...

**Comment by TParcollet** [comment]
Could you elaborate a bit more on what you did to observe again this error ?

**Comment by lizhimao** [comment]
It's the same as before. I set n_chunks for train to 50, but the script run with chunks 51. The 50th config file is the same as the one I send above, in list_chunks.txt you can see, there are 51 train config file, the config file of dev and test also 51. I deleted the exp file before change the chunks, but it seems it still run with chunks+1. Change tarin_with to dev or test, and set n_chunks to 2, it still shows the chunks is 51.

**Comment by mravanelli** [comment]
which python version are you using?

**Comment by lizhimao** [comment]
Python 3.6 with miniconda
and
cuda 9.0
pytorch 1.1.0

**Comment by mravanelli** [comment]
with python 3.7 do you have the same issue?

**Comment by lizhimao** [comment]
Yes. Have you did some configuration before run the code?

**Comment by mravanelli** [comment]
The number of chunks is computed in line 160 of the run_exp.py script. It
calls a function called compute_n_chunks.
Probably doing some prints inside it could help you figure out what is
going on in your case..

**Comment by lizhimao** [comment]
@mravanelli  I found the problem...is quite funny...is only a stupid out_folder name...
Because I set the out folder with ck 50...so the function just take that value, so it always have problem. 
Thank you so much for the advice and patience, also many thank for @TParcollet .

Best

**Comment by TParcollet** [comment]
@lizhimao You're saying that the ck50 in the out_folder is causing this bug ?

**Comment by mravanelli** [comment]
Ok, good to know it. In the next version that will be available quite soon
we will consider it...

Thank you!

Mirco

**Comment by lizhimao** [comment]
Yes, because with print 're.findall('_ck(.+)_', last_ck)[-1]' in compute_n_chunks, I got '50_layer3/exp_files/train_tedlium_tr_ep06_ck49', the following code 're.findall('_ck(.+)_', last_ck)[-1].split('_')[0]' will take the first 50 but not the last 49. 

Last_chunk is:
'last_chunk---- exp/tedlium_MLP_mfcc_ep24_ck50_layer3/exp_files/train_tedlium_tr_ep06_ck49_mfcc.lst'

Thank you again for help!

**Comment by mravanelli** [comment]
You are welcome and thank you for pointing out this!
